dev/tools/controllerbuilder/pkg/codebot/chat.go

```diff
@@ -193,75 +136,80 @@ func (c *Chat) Close() error {
-
-				// diffs don't seem to apply very reliably (????)
-				// if strings.Contains(s, "```diff") {
-				// 	inDiff := false
-				// 	var diff bytes.Buffer
-				// 	for _, line := range strings.Split(s, "\n") {
-				// 		if strings.HasPrefix(line, "```diff") {
-				// 			inDiff = true
-				// 		} else if strings.HasPrefix(line, "```") {
-				// 			inDiff = false
-				// 		} else if inDiff {
-				// 			diff.WriteString(line)
-				// 			diff.WriteString("\n")
-				// 		}
-				// 	}
-
-				// 	cmd := exec.CommandContext(ctx, "patch", "-p1")
-				// 	cmd.Stdin = &diff
-				// 	cmd.Stdout = os.Stdout
-				// 	cmd.Stderr = os.Stderr
-				// 	if err := cmd.Run(); err != nil {
-				// 		fmt.Fprintf(os.Stderr, "PATCH DID NOT APPLY: %v\n", err)
-				// 	} else {
-				// 		fmt.Fprintf(os.Stdout, "Applied patch :-)\n")
-				// 	}
-				// }
-			} else if functionCall, ok := part.(genai.FunctionCall); ok {
-				// klog.Infof("functionCall: %+v", functionCall)
-				klog.Infof("functionCall: %+v", functionCall.Name)
-				response, err := c.runFunctionCall(ctx, functionCall)
-				if err != nil {
-					return fmt.Errorf("unexpected error running function: %w", err)
+				c.ui.AddLLMOutput(&ui.LLMOutput{Text: s})
+				klog.V(2).Infof("TEXT: %+v", s)
+			}
+			if functionCalls, ok := part.AsFunctionCalls(); ok {
+				for _, functionCall := range functionCalls {
+					// klog.Infof("functionCall: %+v", functionCall)
+					klog.V(2).Infof("functionCall: %+v", functionCall.Name)
+					c.ui.AddLLMOutput(&ui.LLMOutput{Text: fmt.Sprintf("functionCall: %+v", functionCall)})
+					result, err := c.runFunctionCall(ctx, functionCall)
+					if err != nil {
+						return fmt.Errorf("unexpected error running function: %w", err)
+					}
+					var response map[string]any
+					if result.Error != nil {
+						response = make(map[string]any)
+						response["result"] = "error"
+						response["error"] = fmt.Sprintf("%v", result.Error)
+						c.ui.AddLLMOutput(&ui.LLMOutput{Text: fmt.Sprintf("error running function: %v", result.Error)})
+					} else {
+						b, err := json.Marshal(result.Response)
+						if err != nil {
+							return fmt.Errorf("converting %T to json: %w", result.Response, err)
+						}
+						m := make(map[string]any)
+						if err := json.Unmarshal(b, &m); err != nil {
+							return fmt.Errorf("unmarshaling json: %w", err)
+						}
+						response = m
+					}
+					functionResponses = append(functionResponses, llm.FunctionCallResult{
+						Name:   functionCall.Name,
+						Result: response,
+					})
+					b, _ := json.Marshal(response)
+					c.ui.AddLLMOutput(&ui.LLMOutput{Text: fmt.Sprintf("sending response: %v", string(b))})
 				}
-				functionResponses = append(functionResponses, genai.FunctionResponse{
-					Name:     functionCall.Name,
-					Response: response,
-				})
-
-			} else {
-				klog.Infof("UNKNOWN: %T %+v", part, part)
 			}
 		}
 
-		if len(functionResponses) != 0 {
-			c.session.History = append(c.session.History, &genai.Content{
```

Comment:
+1. We don't need to record the history

Reply:
Right, some of the client libs do this for us